packages/graalvm/src/main/kotlin/elide/runtime/node/querystring/NodeQuerystring.kt

```diff
@@ -12,36 +12,356 @@
-    ModuleRegistry.deferred(ModuleInfo.of(NodeModuleName.QUERYSTRING)) { singleton }
+    ModuleRegistry.deferred(ModuleInfo.of(NodeModuleName.QUERYSTRING)) { querystring }
   }
 }
 
 /**
  * # Node API: `querystring`
  */
 internal class NodeQuerystring : ReadOnlyProxyObject, QuerystringAPI {
-  //
 
   internal companion object {
+    // Characters that Node.js querystring.escape() encodes beyond non-ASCII and %
+    // These are the characters NOT in the Node.js whitelist: A-Z a-z 0-9 - _ . ! ~ * ' ( )
+    private val additionalEncodeChars = charArrayOf(
+      // RFC 3986 reserved characters that Node.js encodes
+      ' ', ':', '/', '?', '#', '[', ']', '@', '$', '&', '+', ',', ';', '=',
+
+      // Other characters that Node.js encodes
+      '"', '<', '>', '\\', '^', '`', '{', '|', '}',
+    ).map { it.code.toByte() }.toByteArray()
+
+    private const val HEX_RADIX = 16
+
     @JvmStatic fun create(): NodeQuerystring = NodeQuerystring()
   }
 
-  // @TODO not yet implemented
+  // PercentCodec configured for querystring behavior
+  // - Encodes additional characters specified above
+  private val codec = PercentCodec(additionalEncodeChars, true)
+
+  private fun valueToString(value: Value): String {
+    return when {
+      value.isString -> value.asString()
+      value.isNull -> "null"
+      value.isNumber -> value.toString()
+      value.isBoolean -> value.toString()
+      value.hasArrayElements() -> {
+        // Handle arrays - convert to JavaScript-like string representation
+        (0 until value.arraySize).joinToString(",") { index -> valueToString(value.getArrayElement(index)) }
+      }
+      value.hasMembers() -> {
+        sequenceOf("toString", "valueOf")
+          .mapNotNull { methodName ->
+            value.getMember(methodName)
+              ?.takeIf { it.canExecute() }
+              ?.runCatching { execute().asString() }
+              ?.getOrNull()
+          }
+          .firstOrNull()
+          ?: throw JsError.typeError("Cannot convert object to primitive value")
+      }
+      value.isHostObject -> {
+        val hostObject = value.asHostObject<Any>()
+        when (hostObject) {
+          is Array<*> -> {
+            hostObject.joinToString(",") { element -> valueToString(Value.asValue(element)) }
+          }
+          is List<*> -> {
+            hostObject.joinToString(",") { element -> valueToString(Value.asValue(element))}
+          }
+          else -> "[object Object]"
+        }
+      }
+
+      else -> "[object Object]"
+    }
+  }
+
+  override fun getMemberKeys(): Array<String> = moduleMembers
+  override fun getMember(key: String?): Any? = when (key) {
+    QS_DECODE -> ProxyExecutable { args ->
+      when (args.size) {
+        1 -> decode(args[0], null, null, null)
+        2 -> decode(args[0], args[1], null, null)
+        3 -> decode(args[0], args[1], args[2], null)
+        4 -> decode(args[0], args[1], args[2], args[3])
```

Comment:
this can be shortened to something like `decode(args[0], args.getOrNull(1), args.getOrNull(2), args.getOrNull(3))`